Sidorina_Lesson5_task1

ElenaLaufeyson opened the pull request:
Тут будут решения 5 урока.

ElenaLaufeyson committed:
Весь решенный 1 тест.

ElenaLaufeyson committed:
Весь решенный 1 тест.

ElenaLaufeyson committed:
Исправила ошибки в первом уроке (синтаксис и неправильно решенную функцию angleInRadian), а также сделала 3 задания из второго урока.

ElenaLaufeyson committed:
Исправила ошибки из второго урока (сделаны только первые 3 задания).

ElenaLaufeyson committed:
Исправляю ошибки из второго урока...

ElenaLaufeyson committed:
Lesson2_task1
Исправлены ошибки в функциях timeForHalfWay и whichRookThreatens и решена функция rookOrBishopThreatens

ElenaLaufeyson committed:
Lesson2_task1
Сделана функция triangleKind.

ElenaLaufeyson committed:
Lesson2_task2
Сделана функция isNumberHappy.

ElenaLaufeyson committed:
Lesson2_task1
Доделан полностью.

ElenaLaufeyson committed:
Lesson2_task2
Доделала полностью.

ElenaLaufeyson committed:
Lesson3_task1
Решила, но еще не полностью задания 3 урока.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена функция fun digitNumber.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена тривиальная задача.

ElenaLaufeyson committed:
Lesson3_task1
Решила одну сложную задачу.

ElenaLaufeyson committed:
Lesson3_task1
Исправлена проблема с функцией sin и добавлена тестовая функция.

ElenaLaufeyson committed:
Lesson3_task1
Доделала весь третий урок.

ElenaLaufeyson committed:
Lesson4_task1
Начала 4 урок.

ElenaLaufeyson committed:
Lesson3_task1
Исправила функции синуса и косинуса.

ElenaLaufeyson committed:
Lesson4_task1
 Сделано еще несколько функций из 4 урока.

ElenaLaufeyson committed:
Lesson4_task1
Исправила функцию convert.

ElenaLaufeyson committed:
Lesson3_task1
Исправила синус и косинус

ElenaLaufeyson committed:
Lesson4_task1
Доделала 4 урок.

ElenaLaufeyson committed:
Lesson4_task1
Исправила ошибку в функции convertToString.

ElenaLaufeyson committed:
Lesson4_task1
Исправила замечания к функциям convertToString и decimalFromString.

ElenaLaufeyson committed:
Lesson4_task1
Исправила замечания к функции convertToString.

ElenaLaufeyson committed:
Lesson5_task1
Сделала функции dateStrToDigit, dateDigitToStr, bestLongJump

ElenaLaufeyson committed:
Lesson5_task1
Сделала сложную функцию flattenPhoneNumber

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,24 @@ fun main(args: Array<String>) {
+    var index = month.indexOf(parts[1])
+    if (index == -1) return ""
+    var numb = 0
+    var year = 0
```
`numb` и `year` можно сделать `val`, если не инициализировать их 0.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,24 @@ fun main(args: Array<String>) {
+    if (parts.size != 3) return ""
+    var month = listOf<String>("января", "февраля", "марта", "апреля", "мая",
+            "июня", "июля", "августа", "сентября", "октября", "ноября", "декабря")
+    var index = month.indexOf(parts[1])
```
`month` и `index` можно сделать `val`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,24 @@ fun main(args: Array<String>) {
+    if (str.isEmpty()) return ""
+    val parts = str.split(" ")
+    if (parts.size != 3) return ""
+    var month = listOf<String>("января", "февраля", "марта", "апреля", "мая",
```
Тип хранимых значений списка указывать необязательно.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,24 @@ fun main(args: Array<String>) {
+    } catch (e: NumberFormatException) {
+        return ""
+    }
+    return String.format("%02d.%02d.%d", numb, index + 1, year)
```
Неплохо бы ещё проверить, что дата сама по себе корректна.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -69,7 +85,26 @@ fun dateStrToDigit(str: String): String = TODO()
+            "июня", "июля", "августа", "сентября", "октября", "ноября", "декабря")
+    var numb = 0
+    var month = 0
+    var year = 0
```
Все эти переменные можно сделать `val`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -69,7 +85,26 @@ fun dateStrToDigit(str: String): String = TODO()
+    if (digital.isEmpty()) return ""
+    val parts = digital.split(".")
+    if (parts.size != 3) return ""
+    var listmonth = listOf<String>("января", "февраля", "марта", "апреля", "мая",
```
Этот список совпадает со списком из предыдущей функции, лучше сделать этот список общим.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -69,7 +85,26 @@ fun dateStrToDigit(str: String): String = TODO()
+    try {
+        numb = parts[0].toInt()
+        month = parts[1].toInt()
+        if (month !in 1..12) return ""
```
Опять же, неплохо бы проверить корректность даты.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +118,31 @@ fun dateDigitToStr(digital: String): String = TODO()
+    val phonenum1 = phonenum.filter { it == '+' || it in '0'..'9' } //оставляем только плюс и цифры
+    //если отфильтрованная строка не совпадает с изначальной, то
+    // там есть символы, кроме + и цифр, что является ошибкой, а иначе все хорошо
+    if (!phonenum.equals(phonenum1)) return ""
```
Для этой проверки лучше воспользоваться `any` или `all`, а не `filter`.

maximpetrov replied:
Это замечание все ещё в силе

ElenaLaufeyson replied:
Я не вижу в данной ситуации смысла исправлять нормально работающую функцию на функцию с any или all, поскольку считаю вариант решения с filter для себя более простым и понятным. Мне кажется, у каждого задания может быть несколько вариантов решения, а не только тот, который предлагаете Вы, поэтому мое решение тоже имеет место быть.

maximpetrov replied:
Ваше решение не оптимально: Вы сначала создаёте копию строки, а потом ещё и сравниваете две строки, вместо того, чтобы просто, не создавая копию, один раз пройти по строке и проверить её.

ElenaLaufeyson replied:
Зато в моем решении проверяется ГОРАЗДО больше, чем, в принципе, необходимо по заданию, что, как мне кажется, должно вызывать похвалу, а не количество замечаний, растущее в геометрической прогрессии. В задании не сказано, как именно решать задачу, поэтому я все же считаю, что моё решение, хоть и не совсем оптимально, рассматривает очень большое количество вариантов и имеет место быть. Функция filter мне кажется более удобной и понятной, и тут я предпочту все же пользоваться ей.
И, если говорить об оптимальности, то разве оптимально в других функциях вместо того, чтобы один раз объявить переменную var и в цикле просто менять ее значение, переносить объявление переменной в место ее инициализации(в цикл) и делать ее Val, каждый раз заново выделяя и удаляя для неё память?

maximpetrov replied:
Во-первых, я не отрицаю, что Вы проверяете больше, чем большинство Ваших одногруппников, но не больше, чем необходимо по заданию. Во-вторых, количество замечаний если и увеличилось, то уж точно не в геометрической прогрессии: большинство новых замечаний - это замечания к исправлениям, а конкретно к этой функции новых замечаний вообще не появлялось. В-третьих, дело даже не столько в производительности, сколько в том, что Ваше решение выглядит таким образом, как будто Вы узнали про функцию `filter` и она Вам так понравилась, что Вы решили использовать её везде, где нужно и где не нужно. В этом месте её использование не оправдано ни с точки зрения оптимальности, ни с точки зрения понятности: куда более очевидное решение просто пройти по символам строки и проверить каждый на правильность. Сделать это можно как "вручную" в цикле, так и с помощью функций `all` или `any`, которые, кстати, не отличаются по синтаксису от `filter` и очень похожи на неё по принципу действия.
P. S. Оптимальность не страдает при переносе переменных внутрь цикла, поскольку память под них всё равно будет выделена один раз, а вот их область видимости уменьшится до минимально необходимой.

ElenaLaufeyson replied:
Исправила данное замечание, воспользовавшись функцией all, а не filter.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +118,31 @@ fun dateDigitToStr(digital: String): String = TODO()
+    val indexplus = phonenum.indexOf('+') //находим индекс плюса
+    if (indexplus != phonenum.lastIndexOf('+')) return "" //если плюс не один - ошибка
+    //если плюса нет или он в начале(нулевой индекс) - хорошо, иначе ошибка
+    if (indexplus > 0) return ""
```
Раз уж Вы так тщательно проверяете формат, то неплохо бы проверить, что корректность префикса, т.е. что между + и открывающей скобкой есть хотя бы одна цифра.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +118,31 @@ fun dateDigitToStr(digital: String): String = TODO()
+    if (phone.isEmpty()) return ""
+    var phonenum = phone.filter { it != ' ' && it != '-' }
+    val indexopen = phonenum.indexOf('(') //находим индекс открывающейся скобки
+    if (indexopen != -1) { //если открывающаяся скобка есть, то проверяем, что она одна
```
Вы три раза ищете в строке символ, который должен встречаться только один раз. Лучше вынести этот поиск в отдельную функцию.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +154,24 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+fun bestLongJump(jumps: String): Int {
+    if (jumps.isEmpty()) return -1
+    var jump = jumps.filter { it != '%' && it != '-' }
+    jump = jump.trim() //убираем пробелы спереди и сзади
```
Нет большого смысла это делать, всё равно у Вас останутся лишние пробелы внутри, которые и так Вы обрабатываете.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +154,24 @@ fun flattenPhoneNumber(phone: String): String = TODO()
-fun bestLongJump(jumps: String): Int = TODO()
+fun bestLongJump(jumps: String): Int {
+    if (jumps.isEmpty()) return -1
+    var jump = jumps.filter { it != '%' && it != '-' }
```
Можно сделать `val`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +154,24 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    if (jumps.isEmpty()) return -1
+    var jump = jumps.filter { it != '%' && it != '-' }
+    jump = jump.trim() //убираем пробелы спереди и сзади
+    var results = jump.split(" ")
```
Можно сделать `val`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +154,24 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    var answer = -1
+    try {
+        for (element in results) {
+            if (element.length != 0 && element.toInt() > answer)
```
Вместо этого цикла, в принципе, можно обработать список с помощью `filter` и/или `map`, а от результата найти `max()`.

Если оставлять так, то не стоит два раза переводить в `Int`, лучше сохранить в переменную результат.

ElenaLaufeyson committed:
Lesson5_task1
Исправила замечания по выполненным ранее функциям 5 урока.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +65,38 @@ fun main(args: Array<String>) {
+    if (month == 4 || month == 6 || month == 9 || month == 11) {
+        if (numb > 30) return false
+    }
+    if (month == 2) {
```
Здесь лучше `else if`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +135,47 @@ fun dateDigitToStr(digital: String): String = TODO()
+}
+
+///////////////////////////////////////////////////
+fun search(index: Int, str: String, ch: Char): Boolean {
```
Вообще я предполагал, что эта функция будет делать всё, то есть и искать первое вхождение тоже.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +187,35 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    val results = jump.split(" ")
+    if (results.size == 0) return -1
+    var answer = -1
+    var elem: Int
```
Объявление этой переменной лучше перенести в место инициализации и сделать `val`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +187,35 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+fun bestLongJump(jumps: String): Int {
+    if (jumps.isEmpty()) return -1
+    var jump = jumps.filter { it != '%' && it != '-' }
+    jump = deletespace(jump)
```
Всё же лучше просто пофильтровать список после разбиения.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +187,35 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+fun deletespace(str: String): String {
+    var new = str.trim()
+    while (new.contains("  "))
+        new = new.replace("  ", " ")
```
Цикл здесь не нужен, `replace` и так заменит все вхождения.

ElenaLaufeyson committed:
Lesson5_task1
Исправила замечания к сделанным функциям

ElenaLaufeyson committed:
Lesson5_task1
 Исправила функции с переводом даты(добавила, что год может быть 0) и решила еще несколько заданий.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -250,7 +284,20 @@ fun plusMinus(expression: String): Int = TODO()
+    if (str.isEmpty()) return -1
+    var strin = str
+    var index = 0
+    strin = strin.trim()
```
Отбрасывание начальных и конечных пробелов здесь лишнее: задание подразумевает, что входная строка корректна, то есть пробелов в начале и конце нет. Если бы они были, то отбрасывать их было бы ошибкой, потому что тогда у Вас сместились бы индексы.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -250,7 +284,20 @@ fun plusMinus(expression: String): Int = TODO()
-fun firstDuplicateIndex(str: String): Int = TODO()
+fun firstDuplicateIndex(str: String): Int {
+    if (str.isEmpty()) return -1
+    var strin = str
```
Переменную можно сделать `val`, если прямо здесь выполнить все нужные преобразования над `str`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -239,7 +255,25 @@ fun bestHighJump(jumps: String): Int = TODO()
+    if (expression.isEmpty()) throw IllegalArgumentException()
+    var express = expression
+    var answer = 0
+    express = deletespace(express)
```
Здесь тоже лучше (красивее и эффективнее) отфильтровать результат `split`, убрав пустые строки, вместо удаления дублирующихся пробелов.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -239,7 +255,25 @@ fun bestHighJump(jumps: String): Int = TODO()
+    try {
+        answer = expressions[0].toInt()
+        for (i in 1..expressions.size - 2 step 2) {
+            if (expressions[i].equals("+")) answer += expressions[i + 1].toInt()
```
Здесь лучше использовать `when`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -228,7 +227,24 @@ fun deletespace(str: String): String {
-fun bestHighJump(jumps: String): Int = TODO()
+fun bestHighJump(jumps: String): Int {
+    if (jumps.isEmpty()) return -1
+    val jump = jumps.filter { it != '%' && it != '-' }
```
Результаты в строке группируются парами, поэтому выкидывать % и - не самое лучшее решение, потому что некоторые пары перестанут быть парами и Вы не сможете перебирать значения парами.

ElenaLaufeyson replied:
Я поняла, что Вы имеете в виду, но не вижу сейчас особого смысла перебирать значения попарно, поскольку в уже написанной мной функции итак учтены все возможные варианты и она работает корректно и без перебирания значений парами.

maximpetrov replied:
Это не совсем так. В такой реализации у Вас никак правильность формата не проверяется. То есть и несколько чисел подряд, и несколько плюсов или других результатов подряд у Вас не вызовут ошибку, хотя должны бы по условию задачи.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -228,7 +227,24 @@ fun deletespace(str: String): String {
+    val jump = jumps.filter { it != '%' && it != '-' }
+    val results = jump.split(" ").filter { it.length != 0 }
+    var answer = -1
+    if (results.size < 2 || results[0] == "+") return -1
```
Эти проверки не нужны, если перебирать значения парами.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -228,7 +227,24 @@ fun deletespace(str: String): String {
+    var answer = -1
+    if (results.size < 2 || results[0] == "+") return -1
+    try {
+        for (i in 0..results.size - 2) {
```
С учётом предыдущих комментариев, здесь лучше перебирать значения попарно, то есть с шагом 2.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -228,7 +227,24 @@ fun deletespace(str: String): String {
+    try {
+        for (i in 0..results.size - 2) {
+            if (results[i] == "+") continue
+            if (results[i + 1] == "+" && results[i].toInt() > answer)
```
Для `results[i].toInt()` лучше завести переменную и не переводить 2 раза.

ElenaLaufeyson committed:
Lesson5_task1
 Исправила замечания по поводу уже выполненных функций и написала функцию mostExpensive.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -142,7 +313,25 @@ fun firstDuplicateIndex(str: String): Int = TODO()
+    val strings = description.filter { it != ';' }.split(" ")
+    if (strings.size < 2) return ""
+    var maxindex = 1
+    var strInt = 0.0
```
Эту переменную можно сделать `val`, если внести её внутрь `for`.

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +258,27 @@ fun bestHighJump(jumps: String): Int = TODO()
-fun plusMinus(expression: String): Int = TODO()
+fun plusMinus(expression: String): Int {
+    if (expression.isEmpty()) throw IllegalArgumentException()
+    val express = expression
```
Эта переменная у Вас теперь лишняя.

ElenaLaufeyson committed:
Lesson5_task1
Исправила замечания.

ElenaLaufeyson committed:
Lesson5_task1
 Исправление функции bestHighJump

ElenaLaufeyson committed:
Lesson5_task1
Исправления...

maximpetrov commented on src/lesson5/task1/Parse.kt:
```diff
@@ -142,7 +314,24 @@ fun firstDuplicateIndex(str: String): Int = TODO()
+        for (i in 1..strings.size - 1 step 2) {
+            val strInt = strings[i].toDouble()
+            if (strInt < 0) return ""
+            if (strInt >= strings[maxindex].toDouble()) {
```
Лучше найденное максимальное значение тоже сохранять, а не только его индекс, чтобы не переводить каждый раз его в double.

ElenaLaufeyson committed:
Lesson5_task1
Исправила замечание к функции mostExpensive.

ElenaLaufeyson committed:
Lesson5_task1
 Исправила замечание к функции flattenPhoneNumber (сделала с any, а не filter).

ElenaLaufeyson committed:
Lesson5_task1
Исправила замечание к функции flattenPhoneNumber (сделала проверку с all*, а не с filter)

ElenaLaufeyson committed:
Lesson6_task1
Начала делать 6 урок(сделала первые две функции)

maximpetrov commented:
Все замечания исправлены, 5 урок принят.
